### DISCUSSION TITLE
Add filter by date, author, room to search

### DISCUSSION THREAD

**Discussion by MurzNN** [element-hq/element-meta#909] [Ideas]
At now search interface allow only search messages by keywords. Will be good to add advanced filter to search form for filter results via other useful fields:
- Date interval (from-to)
- Author of message
- Mentioned user
- Room name & description (for `All rooms` search)

Good implementation of this filter we can see at Twist app: https://static.twistapp.com/8545d0058b88794f6bb55eda07d40003.png

**Comment by MurzNN** [comment]
Seems at now there are no Client-Server API that can filter search results by date interval: https://matrix.org/docs/spec/client_server/r0.2.0.html#id60 so this must be also implemented in server side too, here is issue about this: https://github.com/matrix-org/matrix-doc/issues/969

**Comment by MurzNN** [comment]
Also will be good to add searching by `star` and other tags, when they will be implemented in https://github.com/vector-im/riot-web/issues/2938

**Comment by turt2live** [comment]
It would also be nice to be able to specify particular tags to look in. I often find myself trying to find that one conversation I had with someone, but my options are manually going through ~900 rooms or using the "All Rooms" function with generic keywords (like "sewing machine"). If I could say "only look at Direct Chats", that'd be much easier to find that *one* conversation I know I had with someone.

**Comment by MurzNN** [comment]
Good example of easy interface with filters we can see in Discord:
![](https://i.imgur.com/fJPGTaA.png)

**Comment by Pestdoktor** [comment]
Experimental branch adding search by author:
https://github.com/Pestdoktor/matrix-react-sdk/tree/author-search

Oddities:
- ~~no~~lacking gui
- only works on current room members
- serverside search only (no e2ee support)

I'd be willing to take this further... Should I make a PR so it can be commented on? Or iron out the aforementioned deficiencies first?

**Comment by aaronraimist** [comment]
It probably makes sense to open a draft PR while you work on it

**Comment by DJCrashdummy** [comment]
referencing https://github.com/vector-im/riot-web/issues/2938#issuecomment-629764841, also a filter for reactions would be nice.

**Comment by DJCrashdummy** [comment]
IMHO also filters for files (#9754), links ( vector-im/element-web#8746) perhaps also mentions and/or keywords (#5382) should be merged into a more advanced search... else sooner or later we would end up with a panel for every filter. :unamused: 
beside that, additionally the discussion, whether a panel should just display the current rooms items or globally, is addressed (*for now i would prefer a switch to toggle either or*) and even combined searches are possible and superior to the panel approach (*"i'm looking for a link and given its topic it must have been in one of these 3 rooms."* -> 1 search with 2 filters and you don't have to go through 3 different rooms one after another).

**Comment by DJCrashdummy** [comment]
thinking a little bit further, this could also help to streamline the UI:
- remove the panel-icons from the top right, except the one for the member-panel
- also remove the search-icon from the top right
- add a search-section to the left panel like *direct messages*, *rooms* etc.:
  - with the `+`-icon you'll get to the full blown search interface (or use instead of the `+` a magnifier-icon to "mark it as a special section")
  - and then (similar to *virtual folders* in thunderbird) within the search-section a bunch of pre-/user-defined searches can be saved without bothering anyone.
- perhaps the saved searches can also be (optionally) included into the *shortcuts to recently viewed rooms above the room list*
- and for sure the results of every search should be displayed in the main window and not squeezed into a side panel

as a result all icons on the top right are room-dependent (`settings`, `integrations`, `share` & `members`) and every user can define & save his own repeatedly needed searches/filters, no mater how much.

btw: the majority of my noobs never look at the top right icons anyway, especially since the right panel is closed by default.
- - -

it seems to me that, if this issue is solved properly, it also closes a bunch of other issues with even higher priority as well. :wink:

**Comment by kittykat** [comment]
The foundations for how we could do filtering are being set up as part of [IA 2 search experience](https://github.com/vector-im/element-web/issues/21121)

**Comment by rltas** [comment]
> Good example of easy interface with filters we can see in Discord

Or Mattermost: https://docs.mattermost.com/channels/search-for-messages.html

**Comment by ell1e** [comment]
I think if I enter a user in the full canonical format (that would be `@user@some.home.server`, right?) it should also be automatically taken as a filter for messages that returns all messages that are either written by that person or mentioning that person in the currently viewed community space (or just the room for the cases where I'm not browsing it through a selected space I'm in).

**Comment by thibaultmol** [comment]
Is any progress being made on this? it's honestly one of the main things that's holding me back from switching to Discord bridging instead of Discord directly. Discord's search is surprisingly decent when it comes to filtering ability

**Comment by spatterIight** [comment]
This is badly wanted feature! <3

**Comment by suse-coder** [comment]
Please add this as soon as possible. Would be very nice as search is very often used.

**Comment by byeuji** [comment]
Adding a voice. This is a deeply needed feature, especially with the coming changes to Discord requiring age verification. I'm looking at moving a community with thousands of active users to Matrix, and searching by user across all rooms within our space/server is a critical necessity.

**Comment by Mr-ex777** [comment]
Adding another voice here, thread has been going on for 9+ years, it's about time.

**Comment by suse-coder** [comment]
Yes search seems to be very underwhelming in element
